modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/near/GridNearTxLocal.java

```diff
@@ -1460,17 +1460,24 @@ private boolean enlistWriteEntry(GridCacheContext cacheCtx,
                                 }
                             }
                             else {
-                                old = entry.innerGet(
-                                    null,
-                                    this,
-                                    /*read through*/false,
-                                    /*metrics*/retval,
-                                    /*events*/retval,
-                                    CU.subjectId(this, cctx),
-                                    entryProcessor,
-                                    resolveTaskName(),
-                                    null,
-                                    keepBinary);
+                                cctx.database().checkpointReadLock();
```

Comment:
The "scope" of this lock should include `entry.innerGetVersioned` as well (a few lines above). `innerGetVersioned` is called by the `getAndPut` method and may lead to expiration.

Reply:
I added a new block of lock-release for this.